src/nvim/eval/typval_encode.h

```diff
@@ -14,54 +14,7 @@
 #include "nvim/eval_defs.h"
 #include "nvim/func_attr.h"
 
-/// Type of the stack entry
-typedef enum {
-  kMPConvDict,  ///< Convert dict_T *dictionary.
-  kMPConvList,  ///< Convert list_T *list.
-  kMPConvPairs,  ///< Convert mapping represented as a list_T* of pairs.
-  kMPConvPartial,  ///< Convert partial_T* partial.
-  kMPConvPartialList,  ///< Convert argc/argv pair coming from a partial.
-} MPConvStackValType;
-
-/// Stage at which partial is being converted
-typedef enum {
-  kMPConvPartialArgs,  ///< About to convert arguments.
-  kMPConvPartialSelf,  ///< About to convert self dictionary.
-  kMPConvPartialEnd,  ///< Already converted everything.
-} MPConvPartialStage;
-
-/// Structure representing current VimL to messagepack conversion state
-typedef struct {
-  MPConvStackValType type;  ///< Type of the stack entry.
-  typval_T *tv;  ///< Currently converted typval_T.
-  int saved_copyID;  ///< copyID item used to have.
-  union {
-    struct {
-      dict_T *dict;    ///< Currently converted dictionary.
-      dict_T **dictp;  ///< Location where that dictionary is stored.
-                       ///< Normally it is &.tv->vval.v_dict, but not when
-                       ///< converting partials.
-      hashitem_T *hi;  ///< Currently converted dictionary item.
-      size_t todo;     ///< Amount of items left to process.
-    } d;  ///< State of dictionary conversion.
-    struct {
-      list_T *list;    ///< Currently converted list.
-      listitem_T *li;  ///< Currently converted list item.
-    } l;  ///< State of list or generic mapping conversion.
-    struct {
-      MPConvPartialStage stage;  ///< Stage at which partial is being converted.
-      partial_T *pt;  ///< Currently converted partial.
-    } p;  ///< State of partial conversion.
-    struct {
-      typval_T *arg;    ///< Currently converted argument.
-      typval_T *argv;    ///< Start of the argument list.
-      size_t todo;  ///< Number of items left to process.
-    } a;  ///< State of list or generic mapping conversion.
-  } data;  ///< Data to convert.
-} MPConvStackVal;
-
-/// Stack used to convert VimL values to messagepack.
-typedef kvec_withinit_t(MPConvStackVal, 8) MPConvStack;
+#include "nvim/eval/typval_encode_types.h"
```

Comment:
eval.c includes directly typval_encode.h, and it only requires MPConvStackVal etc types.
So I thought type definitions of MPConvStack* should be separated from typval_encode.h as typval_encode_types.h (should be typval_encode_defs.h), and eval.c should include it directly.
I think this is the way you commented we should go.
> Most likely before this headers will need to be seriously refactored: I suggested an idea somewhere that all types need to be moved to a separate headers …_types.h or …_defs.h (but it should be generally avoided to put non-types there) which may be arranged in dependency graph without cycles and other headers should include only them